airbyte-integrations/connectors/source-snowflake/src/main/java/io.airbyte.integrations.source.snowflake/SnowflakeSource.java

```diff
@@ -7,66 +7,98 @@
 import com.fasterxml.jackson.databind.JsonNode;
 import com.google.common.collect.ImmutableMap;
 import io.airbyte.commons.json.Jsons;
+import io.airbyte.db.jdbc.JdbcDatabase;
+import io.airbyte.db.jdbc.StreamingJdbcDatabase;
 import io.airbyte.integrations.base.IntegrationRunner;
 import io.airbyte.integrations.base.Source;
 import io.airbyte.integrations.source.jdbc.AbstractJdbcSource;
+import java.io.IOException;
 import java.sql.JDBCType;
+import java.sql.SQLException;
 import java.util.Set;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import javax.sql.DataSource;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
 public class SnowflakeSource extends AbstractJdbcSource<JDBCType> implements Source {
 
   private static final Logger LOGGER = LoggerFactory.getLogger(SnowflakeSource.class);
   public static final String DRIVER_CLASS = "net.snowflake.client.jdbc.SnowflakeDriver";
+  public static final ScheduledExecutorService SCHEDULED_EXECUTOR_SERVICE = Executors.newScheduledThreadPool(1);
 
   public SnowflakeSource() {
-    super(DRIVER_CLASS, new SnowflakeJdbcStreamingQueryConfiguration(), new SnowflakeSourceOperations());
+    super(DRIVER_CLASS, new SnowflakeJdbcStreamingQueryConfiguration(),
+        new SnowflakeSourceOperations());
   }
 
   public static void main(final String[] args) throws Exception {
     final Source source = new SnowflakeSource();
     LOGGER.info("starting source: {}", SnowflakeSource.class);
     new IntegrationRunner(source).run(args);
+    SCHEDULED_EXECUTOR_SERVICE.shutdownNow();
     LOGGER.info("completed source: {}", SnowflakeSource.class);
   }
 
+  @Override
+  public JdbcDatabase createDatabase(JsonNode config) throws SQLException {
+    final DataSource dataSource = SnowflakeDataSourceUtils.createDataSource(config);
+    var database = new StreamingJdbcDatabase(dataSource, new SnowflakeSourceOperations(),
+        new SnowflakeJdbcStreamingQueryConfiguration());
+    quoteString = database.getMetaData().getIdentifierQuoteString();
+    return database;
+  }
+
   @Override
   public JsonNode toDatabaseConfig(final JsonNode config) {
+    final String jdbcUrl = SnowflakeDataSourceUtils.buildJDBCUrl(config);
 
-    final StringBuilder jdbcUrl = new StringBuilder(String.format("jdbc:snowflake://%s/?",
-        config.get("host").asText()));
+    if (config.has("credentials") && config.get("credentials").has("auth_type")
```

Comment:
could you somehow simplify all this logic?

Reply:
done